DIPs/DIP1NNN-DK.md

```diff
@@ -0,0 +1,696 @@
+> For other users, you could simply introduce getters and setters, and functions/methods can already be marked unsafe.
+
+[source](https://github.com/rust-lang/rfcs/pull/80#issuecomment-43489000)
+
+Ultimately the proposal has not been accepted yet.
+The idea of using `private` instead of `@system` variables for D will be discussed in the [alternatives](#alternatives) section.
+More information about Rust's stand on unsafe functions can be found here:
+
+- [safe unsafe meaning](https://doc.rust-lang.org/nightly/nomicon/safe-unsafe-meaning.html)
+- [The scope of unsafe](https://www.ralfj.de/blog/2016/01/09/the-scope-of-unsafe.html)
+
+## Description
+
+### Existing rules for `@system`
+
+Before the proposed changes, we quickly go over the relevant existing rules of what declarations can have the `@system` attribute:
+```D
+@system int w = 2; // compiles, does nothing
+@system enum int x = 3; // compiles, does nothing
+enum E {
+    @system x, // error: @system is not a valid attribute for enum members
+    y,
+}
+@system alias x = E; // compiles, does nothing
+@system template T() {} // compiles, does nothing
+
+void func(@system int x) // error: @system attribute for function parameter is not supported
+{
+    @system int x; // compiles, does nothing
+}
+template Temp(@system int x) {} // error: basic type expected, not @
+```
+In short, anything that can be marked `private` can also be marked `@system`.
+Additionally, local variables can be marked `@system` (while they can't be marked `private`).
+
+Any function attribute can be attached to a variable declaration, but they cannot be retrieved:
+```D
+@system @nogc pure nothrow int x;
+pragma(msg, __traits(getFunctionAttributes, x)); // Error: first argument is not a function
+pragma(msg, __traits(getAttributes, x)); // tuple()
+```
+
+### Unsafe types
+In the proposed changes I use the term "unsafe types".
+An unsafe type is a type which has underlying bit-patterns that risk causing memory corruption in `@safe` code.
+The prime example of an unsafe type is a pointer, but any type that contains a pointer is also unsafe:
+- classes
+- arrays
+- associative arrays
+- a struct or union containing a pointer or one of the above types
+
+Note that unsafe types can be used just fine in `@safe` code for the most part, there are only some restrictions to ensure their integrity:
+- they can't be void-initialized
+- they can't be overlapped in a union
+- a `T[]` cannot be cast to a `U[]` when U is an unsafe type.
+- certain operators (pointer arithmetic, unsafe casts) are disallowed
+
+In the proposed changes both the list of unsafe types and the the list of restrictions will be extended.
+
+### Proposed changes
+
+**(0) _Writing_ to variables or fields marked `@system` is not allowed in `@safe` code**
+
+This does not apply to local variables however, where `@system` remains ignored.
```

Comment:
I don't understand the difference in local variables. Users are unlikely to mark these as `@system` unless there's good reason for it, and it's generally not a good idea to introduce exceptions unless necessary.

Reply:
Adding the checks for local variables is more implementation work with no additional benefit, since there is no way local variables are accessed from the outside unless the function allows references to them to escape.

Reply:
I agree with @atilaneves  that it is not a good idea to have exceptions like this. If no such checking is wanted for local variables, I'd suggest to simply disallow an ˋ@systemˋ attribute on a local variable. That would also prevent a possible later introduction of ˋ@systemˋ-checking on local variables from being a breaking change.

Reply:
Disallowing safety attributes would be inconsistent with the fact that you can still mark a local variable `pure` `nothrow` `@nogc` `@property`, and disallowing useless attributes in general is out of scope for this DIP. I don't think breaking code is a concern since there is currently no reason to add function attributes to local variables.

Reply:
Is this a possible use case for local system variables?

```D
void foo()() // note the template for inference
{
   @system int x; // marked system for semantic reasons

   void bar() @safe { ++x; }
}
```

Reply:
Can you give a more concrete example than foo-bar? I do think there are uses for local system variables, but if I were to add them to this DIP, I think I need a compelling argument that they are worth the extra implementation cost.

Reply:
There's a common pattern in template programming in D where you have a template function with inferred attributes, and regardless of how they are inferred, you can mark specific parts `@trusted` or `@safe` to guarantee those bits are properly checked. Making it so a `@safe` function can access the outer stack frame that is `@system` is questionable, especially if you cannot mark data as untouchable.

The stack frame should be treated like a struct that's on the stack. I don't see why a variable on the stack should be any less useful than a struct member. Otherwise, you will have something like this:

```D
void foo()() {
struct Frame {
   @system int x;
}
Frame frame;
...
}
```

Which is kind of annoying.

Also, it should be mentioned that the foo-bar example really isn't the problem. The real problem is this:

```D
void foo()()
{
   @system int x; // thinking: no safe function can touch this
   void bar() @safe {
      ... // doesn't use x, because it's "not allowed"
   }
}
```

And then later someone edits bar to use x and the original developers instructions were silently ignored by the compiler.

> I think I need a compelling argument that they are worth the extra implementation cost

I don't know what the implementation cost is. I'm very ignorant when it comes to that. So maybe it's difficult to do. From a user perspective, I would want the feature to be available. For consistency's sake as well. I don't know how to judge the "cost" of implementing it.

It would be akin to having `@safe` attributes ignored for inner functions.

Reply:
> I don't know what the implementation cost is. I'm very ignorant when it comes to that. So maybe it's difficult to do.

I've started toying with an [implementation](https://github.com/dlang/dmd/compare/master...dkorpel:systemvariables), and it turns out local variables work automatically, and adding an exception for them is actually *more* code. 

I'll change the DIP.